### PR TITLE
Implement RFC #2 - casting hierarchy between JS values

### PR DESCRIPTION
**Pull request by alexcrichton**
This is an implementation of https://github.com/rustwasm/rfcs/pull/2 for the `wasm-bindgen` suite of crates. Nothing too too tricky here and the commits inline should explain what they're doing.

One question to grapple with is that today we generate `From<JsValue> for ImportedType`, but that's not intended from RFC 2, so we'll want to remove it at some point. This PR currently doesn't delete it so we can avoid making a major bump to 0.3, but we may want to go ahead and bump to 0.3 depending on other schedules!

> **Note:** at the time of this writing the RFC isn't merged yet, so this shouldn't be merged until that's done.

### CODE REVIEW

File: src/cast.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        }
+    }
+
+    /// Performs a zero-cost unchecked cast into the specified type.
+    ///
+    /// This method will convert the `self` value to the type `T`, where both
+    /// `self` and `T` are simple wrappers around `JsValue`. This method **does
+    /// not check whether `self` is an instance of `T`**. If used incorrectly
+    /// then this method may cause runtime exceptions in both Rust and JS, this
+    /// shoudl be used with caution.
```
Comment (fitzgen) [inline src/cast.rs]: "shoudl" -> "should"

File: crates/backend/src/codegen.rs [hunk 2]
```diff
@@ -527,84 +529,85 @@ impl ToTokens for ast::ImportType {
                 obj: ::wasm_bindgen::JsValue,
             }
 
-            impl ::wasm_bindgen::describe::WasmDescribe for #name {
-                fn describe() {
-                    ::wasm_bindgen::JsValue::describe();
+            #[allow(bad_style)]
+            const #const_name: () = {
```
Comment (fitzgen) [inline crates/backend/src/codegen.rs]: lol this is a nice/nasty trick

File: crates/backend/src/codegen.rs [hunk 3]
```diff
@@ -594,18 +596,63 @@ impl ToTokens for ast::ImportType {
                     }
                 }
 
+                // TODO: remove this on the next major version
```
Comment (fitzgen) [inline crates/backend/src/codegen.rs]: Can one mark trait impls as deprecated?

File: src/cast.rs (hunk 1 above)
Comment (afdw) [reply src/cast.rs]: Still not fixed in 2 other places.

Comment (alexcrichton) [reply src/cast.rs]: Oops thanks, fixed now.

File: guide/src/design/import-customization.md [hunk 4]
```diff
@@ -168,3 +168,25 @@ possibilities!
 
   All of these functions will call `console.log` in JS, but each identifier
   will have only one signature in Rust.
+
+* `extends = Bar` - this can be used to say that an imported type extends
```
Comment (fitzgen) [inline guide/src/design/import-customization.md]: This is going to need to move to `guide/src/reference/attributes/on-js-imports/extends.md` and a corresponding entry in the SUMMARY.md table of contents

**Comment by alexcrichton** [comment]
Updated!

File: guide/src/design/import-customization.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
+# Customizing import behavior
```
Comment (fitzgen) [inline guide/src/design/import-customization.md]: This file got accidentally resurrected.

Comment (alexcrichton) [reply guide/src/design/import-customization.md]: Oops, should be dead for real now

**Pull request merged**